projects/output-and-error-handling/client/main.go

```diff
@@ -0,0 +1,100 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+const (
+	url        = "http://localhost:8080"
+	maxRetries = 3
+)
+
+func main() {
+
+	// Create an HTTP client
+	client := &http.Client{}
+	// This loop will run only if the number of reties doesn't exceed 3
+	for retries := 0; retries < maxRetries; retries++ {
+		resp, err := getWeather(url, client)
+		if err != nil {
+			handleError(err)
+			os.Exit(1)
+		}
+		// Close the response body after been fully read
+		defer resp.Body.Close()
```

Comment:
FYI `defer`s get called at the end of the _function_, whereas I suspect you really want this to happen at the end of the _loop iteration_. In this instance it doesn't really matter (the program will be done soon enough regardless), but it's good to know e.g. if you're writing a server that loops for a long time you'll end up leaking memory if you `defer` inside a loop.

To fix this, you could pull out a function which does "the contents of the for loop" :)